servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ScanWithLifetimePublisher.java

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.internal.FlowControlUtils;
+import io.servicetalk.concurrent.internal.SignalOffloader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.OnSubscribeIgnoringSubscriberForOffloading.offloadWithDummyOnSubscribe;
+import static io.servicetalk.concurrent.internal.SubscriberUtils.isRequestNValid;
+import static io.servicetalk.concurrent.internal.SubscriberUtils.newExceptionForInvalidRequestN;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicLongFieldUpdater.newUpdater;
+
+final class ScanWithLifetimePublisher<T, R> extends AbstractNoHandleSubscribePublisher<R> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScanWithLifetimePublisher.class);
+
+    private final Publisher<T> original;
+    private final Supplier<? extends ScanWithLifetimeMapper<? super T, ? extends R>> mapperSupplier;
+
+    ScanWithLifetimePublisher(Publisher<T> original,
+                              Supplier<? extends ScanWithLifetimeMapper<? super T, ? extends R>> mapperSupplier,
+                              Executor executor) {
+        super(executor, true);
+        this.mapperSupplier = requireNonNull(mapperSupplier);
+        this.original = original;
+    }
+
+    @Override
+    void handleSubscribe(final Subscriber<? super R> subscriber, final SignalOffloader signalOffloader,
+                         final AsyncContextMap contextMap, final AsyncContextProvider contextProvider) {
+        original.delegateSubscribe(new ScanWithSubscriber<>(subscriber, mapperSupplier.get(), signalOffloader,
+                contextMap, contextProvider), signalOffloader, contextMap, contextProvider);
+    }
+
+    private static final class ScanWithSubscriber<T, R> implements Subscriber<T> {
+        private static final int NOT_ALLOWED = -1;
```

Comment:
Does using an int versus an enum have an advantage other than potentially the size of the field? ie. why not use AtomicReferenceFieldUpdater

Reply:
No benefit having an int in this case, so I think I will go with your suggestion. 🚀

Reply:
fwiw `int` is the smallest unit of memory for atomics, enums are classes/references which may consume the same size as `int` but maybe not.

Reply:
Understood, references could be 64-bit in VMs with very large heaps where compressed OOPs cannot be used. It does feel a bit like premature optimization to use an `int` rather than enum but I can understand not wanting to revisit.